### PR TITLE
Prevents flicker() via Boo! when shadowling is near.

### PR DESCRIPTION
**Pull request by Bone-White**
Proposed change for #5786 

Allows Ghosts to cast boo but prevents the flickering of lights when a
Shadowling is near.  The Ghost is _not_ informed that their flicker
hasn't worked because they may not have Antag HUD enabled.

N.B. range parameter on line 32 of range() may need adjusting as I have
approximated it's effectiveness.

### CODE REVIEW

**Comment by FlattestGuitar** [comment]
Ghost metagame is a thing now. I see...

**Comment by Plotron** [comment]
Well. If it's a hatched shadowling, then you don't really need an ANTAG HUD to know it's an antag. RIGHT?

File: code/modules/mob/dead/observer/spells.dm [hunk 1]
```diff
@@ -26,6 +26,13 @@ var/global/list/boo_phrases=list(
 	range = 1 // Or maybe 3?
 
 /obj/effect/proc_holder/spell/aoe_turf/boo/cast(list/targets)
+
+	// First check if there are shadowlings nearby, because flickering lights will damage them.
+	var/shadowling_present = 0
+	for(var/mob/living/carbon/C in range(7, usr)) // (Range is Ghost Boo range (1) + Shadowling Veil range (5) + 1 for good measure)
+		if (istype(C,/datum/species/shadow/ling))
```
Comment (Fox-McCloud) [inline code/modules/mob/dead/observer/spells.dm]: This check won't work; a carbon can't be a species datum.

You'd want to cycle through humans in range and you'd want to check with `get_species`

**Comment by Bone-White** [comment]
@Plotron You're right, I'll add some flavour text to it's use when near Shadowlings

@Fox-McCloud Line 33 was supposed to be C.species, but if there's a proc that handles it I'll use that.  I wish there was a directory of SS13 procs, I guess all I can rely on is your feedback on PRs, so thanks.

Interactions between two clients at the same time are difficult to test, so even though my code was wrong it appeared to work as I expected when I tested it ho hum.

Updated PR incoming.

**Comment by Bone-White** [comment]
That should do it, I tested this more thoroughly this time.

File: code/modules/mob/dead/observer/spells.dm [hunk 2]
```diff
@@ -59,4 +63,7 @@ var/global/list/boo_phrases=list(
 				A:spookymode=1
 
 			if(istype(A, /obj/machinery/ai_status_display))
-				A:spookymode=1
\ No newline at end of file
+				A:spookymode=1
+
+	if (shadowling_blocked_light_flicker)
+		to_chat(action.owner.client, "The nearby shadowling laughs at your unsuccessful attempt to flicker lights in it's presence.")
```
Comment (Plotron) [inline code/modules/mob/dead/observer/spells.dm]: http://www.elearnenglishlanguage.com/blog/english-mistakes/its/
Grammar. It's vs its.

**Comment by Bone-White** [comment]
Fixes #5792

File: code/modules/mob/dead/observer/spells.dm [hunk 3]
```diff
@@ -26,6 +26,16 @@ var/global/list/boo_phrases=list(
 	range = 1 // Or maybe 3?
 
 /obj/effect/proc_holder/spell/aoe_turf/boo/cast(list/targets)
+
+	// First check if there are shadowlings nearby, because flickering lights will damage them.
+	var/shadowling_present = 0
+	for(var/mob/living/carbon/C in range(7, usr)) // (Range is Ghost Boo range (1) + Shadowling Veil range (5) + 1 for good measure)
+		if (C.get_species() == "Shadowling")
```
Comment (Fox-McCloud) [inline code/modules/mob/dead/observer/spells.dm]: I recommend checking for Shadowling, Lesser Shadowling (yes, that's a thing), and Shadow people. 

so `Shadowling`, `Lesser Shadowling`, and `Shadow`.

File: code/modules/mob/dead/observer/spells.dm [hunk 4]
```diff
@@ -49,4 +63,7 @@ var/global/list/boo_phrases=list(
 				A:spookymode=1
 
 			if(istype(A, /obj/machinery/ai_status_display))
-				A:spookymode=1
\ No newline at end of file
+				A:spookymode=1
+
+	if (shadowling_blocked_light_flicker)
+		to_chat(action.owner.client, "The nearby shadowling laughs at your unsuccessful attempt to flicker lights in their presence.")
```
Comment (Fox-McCloud) [inline code/modules/mob/dead/observer/spells.dm]: I'd just make this fail silently.

File: code/modules/mob/dead/observer/spells.dm [hunk 5]
```diff
@@ -49,4 +63,7 @@ var/global/list/boo_phrases=list(
 				A:spookymode=1
 
 			if(istype(A, /obj/machinery/ai_status_display))
-				A:spookymode=1
\ No newline at end of file
+				A:spookymode=1
+
+	if (shadowling_blocked_light_flicker)
```
Comment (Fox-McCloud) [inline code/modules/mob/dead/observer/spells.dm]: no space after `if` please.

File: code/modules/mob/dead/observer/spells.dm [hunk 6]
```diff
@@ -37,9 +47,13 @@ var/global/list/boo_phrases=list(
 
 			// Flicker unblessed lights in range
 			if(istype(A,/obj/machinery/light))
-				var/obj/machinery/light/L = A
-				if(L)
-					L.flicker()
+				if(! shadowling_present)  // If no shadowling is nearby
```
Comment (Fox-McCloud) [inline code/modules/mob/dead/observer/spells.dm]: `if(!shadowling_present)`

**Comment by Bone-White** [comment]
I'm wondering if there's a style manual you have for code, for things like "no spaces after if".

Also, you'll want to check my get_species() if statement, I doubt that will match the expected style either.

**Comment by KingPhilipIII** [comment]
I've actually seen boos used to rescue shadowlings as well. It was locked down and it's veil and shadow walk must have been recharging because it was burning to death. Removed a flickering bulb before being critted

**Comment by Markolie** [comment]
Looks good to me.

**Comment by Fox-McCloud** [comment]
This won't actually resolve the issue, because, well, if it's flickering before the shadowlings casts veil, it'll just continue to flicker---likewise, how this is addressed is veryyy snowflakey.

**Comment by Bone-White** [comment]
Well, you've got various ways of dealing with this, care to make a suggestion?

**Comment by Fox-McCloud** [comment]
Basically, you need to prevent `flicker` from turning lights on when they're already off, or having it so that shadowling veil just out and out ends `flicker`, even if `flicker` is only midway through.

**Comment by Fox-McCloud** [comment]
You plan on working on this further?

**Comment by Fox-McCloud** [comment]
No response--closing this. If you happen to come up with something further, feel free to make a new PR or re-open this one.

**Pull request closed without merging**